cdap-runtime-ext-dataproc/src/main/java/io/cdap/cdap/runtime/spi/common/ArtifactCacheManager.java

```diff
@@ -0,0 +1,182 @@
+      Set<String> files = getCachedArtifactsForRun(client, bucket, cachedArtifactsListFilePath);
+      String cachedArtifactsPath = getPath(DataprocUtils.CDAP_GCS_ROOT, DataprocUtils.CDAP_CACHED_ARTIFACTS);
+      String cacheCountFilePath = getPath(cachedArtifactsPath, CACHED_ARTIFACTS_USAGE_COUNT);
+      changeUsageCountForArtifacts(client, bucket, files, cachedArtifactsPath, cacheCountFilePath, -1);
+    } catch (Exception e) {
+      LOG.error("Error in releaseCacheUsageForArtifacts", e);
+    }
+  }
+
+  private static void changeUsageCountForArtifacts(Storage client, String bucket, Set<String> files, String cachePath,
+                                                   String filePath, int changeValue) {
+    if (files.isEmpty()) {
+      LOG.debug("No Cached Artifacts found for this run!");
+      return;
+    }
+    BlobId blobId = BlobId.of(bucket, filePath);
+    for (int i = 0; i < MAX_RETRIES_FOR_CACHE_COUNTER_OPERATION; i++) {
+      try {
+        Blob blob = client.get(blobId);
+        Map<String, Integer> artifactCount;
+        if (blob == null || !blob.exists()) {
+          LOG.debug("Creating ArtifactsCacheCounter {} at {}", CACHED_ARTIFACTS_USAGE_COUNT, filePath);
+          artifactCount = new HashMap<>();
+          BlobInfo createBlob = BlobInfo.newBuilder(blobId).setContentType(CONTENT_TYPE_JSON).build();
+          try {
+            client.create(createBlob, GSON.toJson(artifactCount).getBytes(StandardCharsets.UTF_8),
+                          Storage.BlobTargetOption.doesNotExist());
+          } catch (StorageException e) {
+            if (e.getCode() != HttpURLConnection.HTTP_PRECON_FAILED) {
+              throw e;
+            }
+          }
+          blob = client.get(blobId);
+        }
+        artifactCount = GSON.fromJson(new String(blob.getContent(), StandardCharsets.UTF_8),
+                                      new TypeToken<Map<String, Integer>>() {
+                                      }.getType());
+        for (String file : files) {
+          int count = artifactCount.getOrDefault(file, 0) + changeValue;
+          if (count <= 0) {
+            if (count < 0) {
+              LOG.error("Cache usage count less than 0 for {} in {}", file, getPath(cachePath,
+                                                                                    CACHED_ARTIFACTS_USAGE_COUNT));
+            }
+            artifactCount.remove(file);
+            setCustomTimeOnArtifactAndReleaseHold(client, bucket, file, cachePath);
+          } else {
+            artifactCount.put(file, count);
+          }
+        }
+        try {
+          WritableByteChannel writer = blob.writer(Storage.BlobWriteOption.generationMatch());
+          writer.write(ByteBuffer.wrap(GSON.toJson(artifactCount).getBytes(StandardCharsets.UTF_8)));
+          writer.close();
+        } catch (StorageException | IOException e) {
+          LOG.error("Exception while writing to artifacts cache counter file", e);
```

Comment:
Shouldn't we handle HttpURLConnection.HTTP_PRECON_FAILED exception here? This is the place where another thread could have changed the counter.

Reply:
If we get a PreCondition exception here, we need to redo the entire thing. It has been handled as such.